nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp

```diff
@@ -408,28 +424,38 @@ bool RegulatedPurePursuitController::isCollisionImminent(
   curr_pose.y = robot_pose.pose.position.y;
   curr_pose.theta = tf2::getYaw(robot_pose.pose.orientation);
 
-  int i = 1;
-  while (true) {
-    // only forward simulate within time requested
-    if (i * projection_time > max_allowed_time_to_collision_) {
-      break;
-    }
+  // compute distace to goal to stop collision checking when closer to goal
+  double dx = global_plan_.poses.back().pose.position.x - robot_pose.pose.position.x;
+  double dy = global_plan_.poses.back().pose.position.y - robot_pose.pose.position.y;
+  double dist_to_goal = std::hypot(dx, dy);
+
 
+  // only forward simulate within time requested
+  int i = 1;
+  while (i * projection_time < max_allowed_time_to_collision_) {
     i++;
 
     // apply velocity at curr_pose over distance
     curr_pose.x += projection_time * (linear_vel * cos(curr_pose.theta));
     curr_pose.y += projection_time * (linear_vel * sin(curr_pose.theta));
     curr_pose.theta += projection_time * angular_vel;
 
+    dx = global_plan_.poses.back().pose.position.x - curr_pose.x;
+    dy = global_plan_.poses.back().pose.position.y - curr_pose.y;
+    dist_to_goal = std::hypot(dx, dy);
+
+    if (dist_to_goal < costmap_->getResolution()) {
```

Comment:
Before in our reviews, you had a condition like `dist_to_goal < desired_linear_vel_ * max_allowed_time_to_collision_` based on velocity and time rather than distance. I think your initial intuition was the right way to go here, if I may explain.

If the current pose projected forward by the current velocity doesn't actually intersect with the goal pose, then this condition would never be hit. I think that's why you saw changes when you made this 2x and 3x the costmap resolution so that if there's some error of the  general region, it hits it OK. 

However, if you measure the distance to the goal from the `curr_pose` and then compare it to the current `projection_time * velocity`, what you're obtaining is the distance traversed on the current heading as compared to the distance to the goal. That would make it so that even if the current velocity didn't pass through the goal pose, we'd still terminate after the theoretical minimum distance to achieve the goal.